src/loadsave_gui.cpp

```diff
@@ -32,12 +63,14 @@ class LoadSaveGui : public GuiWindow {
 	void DrawWidget(WidgetNumber wid_num, const BaseWidget *wid) const override;
 	void OnClick(WidgetNumber wid, const Point16 &pos) override;
 	bool OnKeyEvent(WmKeyCode key_code, WmKeyMod mod, const std::string &symbol) override;
+	void Sort(const SortBy &sort);
 
 private:
 	std::string FinalFilename() const;
 
-	const Type type;                  ///< Type of this window.
-	std::set<PreloadData> all_files;  ///< All files in the working directory.
+	const Type type;                   ///< Type of this window.
+	std::list<PreloadData> all_files;  ///< All files in the working directory.
+	const SortBy *current_sort;        ///< Current sorting order.
```

Comment:
Avoiding std::function does mean you have to use void* to store the function pointer which isn't so nice for the typesystem, but suitable amount of templating (on Sort) should get you pretty close